### PR TITLE
Implement Subresource Integrity

### LINKED ISSUE #14523: Implement Subresource Integrity (SRI)
https://w3c.github.io/webappsec-subresource-integrity/

### PR DESCRIPTION
**Pull request by nmvk**
Implemented response validation part of https://w3c.github.io/webappsec-subresource-integrity/.
Implemented step eighteen of the main fetch. If a request has integrity metadata, then following steps are performed
1) Wait for response body
2) If the response does not have a termination reason and response does not match request’s integrity metadata, set response and internalResponse to a network error.

Dependency updated: html5ever-atoms from 0.1.2 to 0.1.3. This will not completely fix #14523, It will implement changes related to response validation. Request validation algorithm implementation needs CSP.

I did not update any WPT-Test. In my local system, I found some assertion issue dependent on the order of execution of test-case. It would be helpful if someone could do "try" build on these changes to get wpt results.

r? @jdm 

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors

- [X] There are tests for these changes 

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14865)

### CODE REVIEW

**Comment by jdm** [comment]
```
  ▶ TIMEOUT [expected OK] /subresource-integrity/subresource-integrity.sub.html
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  └ 3.3 (Core Profile) Mesa 12.0.1

  ▶ Unexpected subtest result in /subresource-integrity/subresource-integrity.sub.html:
  └ NOTRUN [expected PASS] Style: Same-origin with correct sha256 and sha512 hash, rel='alternate stylesheet' enabled

  ▶ Unexpected subtest result in /subresource-integrity/subresource-integrity.sub.html:
  └ NOTRUN [expected PASS] Style: Same-origin with incorrect sha256 and sha512 hash, rel='alternate stylesheet' enabled

  ▶ Unexpected subtest result in /subresource-integrity/subresource-integrity.sub.html:
  │ FAIL [expected PASS] Style: <crossorigin='anonymous'> with correct hash, ACAO: *
  │   → assert_unreached: Good load fired error handler. Reached unreachable code
  │ FAIL [expected PASS] Style: <crossorigin='use-credentials'> with correct hash, CORS-eligible
  │   → assert_unreached: Good load fired error handler. Reached unreachable code
  │ 
  │ SRIStyleTest.prototype.execute/</<@http://web-platform.test:8000/subresource-integrity/subresource-integrity.sub.html:132:39
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  └ SRIStyleTest.prototype.execute/<@http://web-platform.test:8000/subresource-integrity/subresource-integrity.sub.html:132:17

  ▶ Unexpected subtest result in /subresource-integrity/subresource-integrity.sub.html:
  │ FAIL [expected PASS] Script: <crossorigin='use-credentials'> with correct hash, CORS-eligible
  │   → assert_unreached: Good load fired error handler. Reached unreachable code
  │ 
  │ SRIScriptTest.prototype.execute/</<@http://web-platform.test:8000/subresource-integrity/subresource-integrity.sub.html:59:39
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  └ SRIScriptTest.prototype.execute/<@http://web-platform.test:8000/subresource-integrity/subresource-integrity.sub.html:59:17

  ▶ Unexpected subtest result in /subresource-integrity/subresource-integrity.sub.html:
  │ FAIL [expected PASS] Style: Same-origin with incorrect hash.
  │   → assert_not_equals: got disallowed value "rgb(255, 255, 0)"
  │ FAIL [expected PASS] Style: Same-origin with sha256 match, sha512 mismatch
  │   → assert_not_equals: got disallowed value "rgb(255, 255, 0)"
  │ FAIL [expected PASS] Style: <crossorigin='anonymous'> with incorrect hash, ACAO: *
  │   → assert_not_equals: got disallowed value "rgb(255, 255, 0)"
  │ FAIL [expected PASS] Style: <crossorigin='use-credentials'> with incorrect hash CORS-eligible
  │   → assert_not_equals: got disallowed value "rgb(255, 255, 0)"
  │ FAIL [expected PASS] Style: <crossorigin='anonymous'> with CORS-ineligible resource
  │   → assert_not_equals: got disallowed value "rgb(255, 255, 0)"
  │ FAIL [expected PASS] Style: Cross-origin, not CORS request, with correct hash
  │   → assert_not_equals: got disallowed value "rgb(255, 255, 0)"
  │ FAIL [expected PASS] Style: Cross-origin, not CORS request, with hash mismatch
  │   → assert_not_equals: got disallowed value "rgb(255, 255, 0)"
  │ 
  │ SRIStyleTest.prototype.execute/</<@http://web-platform.test:8000/subresource-integrity/subresource-integrity.sub.html:141:21
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  └ SRIStyleTest.prototype.execute/<@http://web-platform.test:8000/subresource-integrity/subresource-integrity.sub.html:139:17
```

**Comment by jdm** [comment]
Let me know if you want me to start reviewing the changes or whether I should wait until you look into the test failure.

**Comment by mrnayak** [comment]
Thanks Josh, I will work on these test cases today. Will notify you once these gets fixed.

**Comment by mrnayak** [comment]
None of the test case failures seem to be related to changes.

1) assertion error in 7 test case is due to stylesheet already been loaded in other success test case. I moved all test case expecting assert_not_equals to execute before any success test case.
2) crossorigin has not been implemented in htmllinkelement. So i changed assertion of two test case (`Style: <crossorigin='anonymous'> with correct hash, ACAO: *` and `Style: <crossorigin='use-credentials'> with correct hash, CORS-eligible`) to false.
3) Two alternate stylesheet related tests got timed out since we ignore alternate stylesheet. I have filed an issue #14881. Currently, I have commented these tests.
4) `Script: <crossorigin='use-credentials'> with correct hash, CORS-eligible` failed due to bug in step six of cors_check of `http_loader` . This has been fixed.

@jdm If my above analysis is correct, can you please review the changes.

**Comment by jdm** [comment]
Rather than modifying the test file, we should be modifying `tests/wpt/metadata/subresource-integrity/subresource-integrity.sub.html.ini` to include the expected failing test results.

**Comment by jdm** [comment]
Can you explain point 1 in more detail? Why does the order make a difference in the test result?

**Comment by mrnayak** [comment]
@jdm I am not sure if point 1 is a bug. Our assertion is to validate that div style background color is not yellow.  Moment first success test case executes stylesheet gets loaded and subsequent all assertion fail though we don't load stylesheet again. 

I update `tests/wpt/metadata/subresource-integrity/subresource-integrity.sub.html.ini`, Thanks

**Comment by jdm** [comment]
I know what the problem is - the test relies on removing the stylesheet from any previous test before starting the new test. However, Servo doesn't support removing stylesheets from the document yet (see #976).

**Comment by mrnayak** [comment]
So I will modify `subresource-integrity.sub.html.ini` to include these tests as well?

**Comment by jdm** [comment]
Yes please!

**Comment by mrnayak** [comment]
@jdm I have done the changes

File: tests/unit/net/fetch.rs [hunk 1]
```diff
@@ -556,6 +555,57 @@ fn test_fetch_with_hsts() {
...
+        request.local_urls_only = false;
+
+        fetch(request, None)
+    };
+
+    let response = do_fetch(server_url);
+
+    let _ = server.close();
+    assert!(response.is_network_error());
+    assert!(response.internal_response.unwrap().is_network_error());
```
Comment (jdm) [inline tests/unit/net/fetch.rs]: This assertion won't pass after previous changes I requested. We should be able to remove it.

File: tests/unit/net/fetch.rs [hunk 2]
```diff
@@ -556,6 +555,57 @@ fn test_fetch_with_hsts() {
...
+        response.send(MESSAGE).unwrap();
+    };
+    let (mut server, server_url) = make_server(handler);
+
+    let do_fetch = |url: ServoUrl| {
+        let origin = Origin::Origin(url.origin());
+        let mut request = Request::new(url, Some(origin), false, None);
+        *request.referrer.borrow_mut() = Referrer::NoReferrer;
+        *request.integrity_metadata.borrow_mut()
+             = "sha384-H8BRh8j48O9oYatfu5AZzq6A9RINhZO5H16dQZngK7T62em8MUt1FLm52t+eX6xO".to_owned();
```
Comment (jdm) [inline tests/unit/net/fetch.rs]: Let's add a comment about how this was generated.

File: tests/unit/net/fetch.rs [hunk 3]
```diff
@@ -556,6 +555,57 @@ fn test_fetch_with_hsts() {
...
+
+#[test]
+fn test_fetch_with_sri_network_error() {
+    static MESSAGE: &'static [u8] = b"alert('Hello, Network Error');";
+    let handler = move |_: HyperRequest, response: HyperResponse| {
+        response.send(MESSAGE).unwrap();
+    };
+    let (mut server, server_url) = make_server(handler);
+
+    let do_fetch = |url: ServoUrl| {
```
Comment (jdm) [inline tests/unit/net/fetch.rs]: This code can be inline instead of using a closure.

File: tests/unit/net/subresource_integrity.rs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    algorithm = get_prioritized_hash_function("sha256", "sha384");
+    assert_eq!(algorithm, "sha384".to_owned());
+
+    algorithm = get_prioritized_hash_function("sha384", "sha512");
+    assert_eq!(algorithm, "sha512".to_owned());
+}
+
+#[test]
+fn test_parsed_metadata_without_options() {
```
Comment (jdm) [inline tests/unit/net/subresource_integrity.rs]: Let's add a test for malformed integrity attributes, too. No `-`s, multiple `?`s, etc.

File: components/net/subresource_integrity.rs [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use net_traits::response::{Response, ResponseBody, ResponseType};
+use openssl::crypto::hash::{hash, Type as MessageDigest};
+use rustc_serialize::base64::{STANDARD, ToBase64};
+use std::collections::HashMap;
+use std::sync::MutexGuard;
+lazy_static! {
+    //Key is supported algorithm and value is priority
+    static ref SUPPORTED_ALGORITHM: HashMap<&'static str, usize> = {
```
Comment (jdm) [inline components/net/subresource_integrity.rs]: We can do this instead:
```rust
const SUPPORTED_ALGORITHM: &'static [&'static str] = [
    "sha256",
    "sha384",
    "sha512",
];
```
and use the position in the array as the priority instead.

File: components/net/subresource_integrity.rs [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+        }
+    }
+
+    return result;
+}
+
+/// https://w3c.github.io/webappsec-subresource-integrity/#getprioritizedhashfunction
+pub fn get_prioritized_hash_function(hash_func_left: &str, hash_func_right: &str) -> String {
+    let left_priority = SUPPORTED_ALGORITHM.get(hash_func_left).unwrap();
+    let right_priority = SUPPORTED_ALGORITHM.get(hash_func_right).unwrap();
```
Comment (jdm) [inline components/net/subresource_integrity.rs]: These can use `SUPPORTED_ALGORITHM.position(|s| s == hash_func_left).unwrap()` instead.

File: components/net/subresource_integrity.rs [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+                result.push(SriEntry::new(alg, digest, opt))
+            }
+        }
+    }
+
+    return result;
+}
+
+/// https://w3c.github.io/webappsec-subresource-integrity/#getprioritizedhashfunction
+pub fn get_prioritized_hash_function(hash_func_left: &str, hash_func_right: &str) -> String {
```
Comment (jdm) [inline components/net/subresource_integrity.rs]: Let's return `Option<&str>` here instead, which will avoid the magic empty string behaviour.

File: components/net/subresource_integrity.rs [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+    }
+
+}
+
+/// https://w3c.github.io/webappsec-subresource-integrity/#get-the-strongest-metadata
+pub fn get_strongest_metadata(integrity_metadata_list: Vec<SriEntry>) -> Vec<SriEntry> {
+    let mut result: Vec<SriEntry> = vec![integrity_metadata_list[0].clone()];
+    let mut current_algorithm = result[0].alg.clone();
+
+    for i in 1..integrity_metadata_list.len() {
```
Comment (jdm) [inline components/net/subresource_integrity.rs]:
```rust
for integrity_metadata in &integrity_metadata_list[1..] {
```

File: components/net/subresource_integrity.rs [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+        let intgerity_metadata = integrity_metadata_list[i].clone();
+
+        let prioritized_hash = get_prioritized_hash_function(&intgerity_metadata.alg,
+                                                                 &*current_algorithm);
```
Comment (jdm) [inline components/net/subresource_integrity.rs]: nit: indentation.

File: components/net/subresource_integrity.rs [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+
+/// https://w3c.github.io/webappsec-subresource-integrity/#apply-algorithm-to-response
+fn apply_algorithm_to_response(body: MutexGuard<ResponseBody>,
+                               message_digest: MessageDigest)
+                               -> String {
+    if let ResponseBody::Done(ref vec) = *body {
+        let response_digest = hash(message_digest, vec);
+        response_digest.to_base64(STANDARD)
+    } else {
+        "".to_owned()
```
Comment (jdm) [inline components/net/subresource_integrity.rs]: We should probably make this `unreachable!("Tried to calculate digest of incomplete response body")` instead.

File: components/net/subresource_integrity.rs [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+        _ => false,
+    }
+}
+
+/// https://w3c.github.io/webappsec-subresource-integrity/#does-response-match-metadatalist
+pub fn is_response_integrity_valid(integrity_metadata: &str, response: &Response) -> bool {
+    let parsed_metadata_list: Vec<SriEntry> = parsed_metadata(integrity_metadata);
+
+    // Step 2 & 4
+    if parsed_metadata_list.len() == 0 {
```
Comment (jdm) [inline components/net/subresource_integrity.rs]: `if parse_metadata_list.is_empty() {`

**Review by jdm — CHANGES_REQUESTED** [review]
This is really great work! I have a number of small changes, but in general I like the way the code is organized and found the changes easy to read.

**Comment by mrnayak** [comment]
Thanks @jdm I have made changes. Please let me know if there are any other changes.

**Comment by jdm** [comment]
@bors-servo: r+
Wonderful!

**Comment by mrnayak** [comment]
I had missed updating expectation of two test case :(. I have corrected this.

**Comment by jdm** [comment]
You will also need to adjust the expectations for /html/dom/interfaces.html, /html/dom/reflection-misc.html, and /html/dom/reflection-metadata.html. I recommend running
`./mach test-wpt tests/wpt/web-platform-test/path/to/test.html --no-pause-after-test --log-raw /tmp/servo.log && ./mach update-wpt /tmp/servo.log` for each of those tests.

**Comment by mrnayak** [comment]
@jdm Thanks, I have fixed expectations and merge conflicts

**Comment by mrnayak** [comment]
Infra issue?

**Pull request merged**